File: src/Other/BestMeetingPoint.java
package Other;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

// https://leetcode.ca/2016-09-21-296-Best-Meeting-Point/
public class BestMeetingPoint {
    public int minTotalDistance(int[][] grid) {

        // empty, null check
        List<Integer> ipos = new ArrayList<>();
        List<Integer> jpos = new ArrayList<>();

        for (int i = 0; i < grid.length; i++) {
            for (int j = 0; j < grid[0].length; j++) {
                if (grid[i][j] == 1) {
                    ipos.add(i); // i is sorted in nature
                    jpos.add(j);
                }
            }
        }

        int sum = 0;
        // median is the min to all the nodes
        int iMid = ipos.get(ipos.size() / 2); // find the median i
        for(Integer pos : ipos){
            sum += Math.abs(pos - iMid);
        }

        Collections.sort(jpos); // sort j
        int jMid = jpos.get(jpos.size() / 2); // find the median of j
        for(Integer pos : jpos){
            sum += Math.abs(pos - jMid);
        }
        // i + j
        return sum;
    }

    public static void main(String[] args) {
        BestMeetingPoint out = new BestMeetingPoint();
        System.out.println(out.minTotalDistance(new int[][] { {1,0,0,0,1},{0,0,0,0,0},{0,0,1,0,0} }));
    }
}

File: py/essential/BestMeetingPoint.py
# https://leetcode.ca/2016-09-21-296-Best-Meeting-Point/

from collections import deque
from math import inf
from typing import List


class Solution:
    def minTotalDistance(self, grid: List[List[int]]) -> int:
        m, n = len(grid), len(grid[0])
        q = deque()
        dist = [[0] * n for _ in range(m)]
        dirs = [(1,0),(0,1),(-1,0),(0,-1)]
        for i in range(m):
            for j in range(n):
                if grid[i][j]==1:
                    q.append((i,j))
                    distance = 0
                    visited = set([(i,j)])
                    while q:
                        distance+=1
                        for _ in range(len(q)):
                            r, c = q.popleft()
                            for dir in dirs:
                                rn, cn = r+dir[0], c+dir[1]
                                if 0<=rn<m and 0<=cn<n and (rn,cn) not in visited:
                                    dist[rn][cn]+=distance
                                    q.append((rn, cn))
                                    visited.add((rn,cn))
        res = inf
        for i in range(m):
            for j in range(n):
                res = min(res, dist[i][j])
        return res

s=Solution()
print(s.minTotalDistance([[1,0,0,0,1],[0,0,0,0,0],[0,0,1,0,0]]))
print(s.minTotalDistance([[1,1]]))